README.md
# @fizzmod/utils

Funciones utiles para desarrollar un ecommerce en vtex

## Install

```bash
npm install --save @fizzmod/utils
```

## Peer Dependecies
- [axios](https://github.com/axios/axios)
- [@babel/runtime](https://babeljs.io/docs/en/babel-runtime)

<a name="Utils"></a>

## Utils : <code>object</code>
**Kind**: global namespace  

* [Utils](#Utils) : <code>object</code>
    * [.Functions](#Utils.Functions) : <code>object</code>
        * [.setCurrency(currency)](#Utils.Functions.setCurrency)
        * [.formatPrice(number, [thousands], [decimals], [length], [currency])](#Utils.Functions.formatPrice) ⇒ <code>string</code>
        * [.sanitizeString(str, [replace])](#Utils.Functions.sanitizeString) ⇒ <code>string</code>
        * [.getResizedImage(src, width, height)](#Utils.Functions.getResizedImage) ⇒ <code>string</code>
        * [.setCookie(cname, cvalue, [exdays], [isdomain])](#Utils.Functions.setCookie) ⇒ <code>void</code>
        * [.getCookie(cname)](#Utils.Functions.getCookie) ⇒ <code>string</code>
        * [.deleteCookie(cname)](#Utils.Functions.deleteCookie) ⇒ <code>void</code>
        * [.addAnimation(name, callback)](#Utils.Functions.addAnimation)
        * [.setStrLength(str, maxLength)](#Utils.Functions.setStrLength) ⇒
        * [.stripHost(URL)](#Utils.Functions.stripHost) ⇒ <code>string</code>
        * [.detectIE()](#Utils.Functions.detectIE) ⇒ <code>string</code> \| <code>false</code>
        * [.calculatePercentDiscount(listprice, bestprice)](#Utils.Functions.calculatePercentDiscount) ⇒ <code>string</code>
        * [.isGoogleMapLoaded()](#Utils.Functions.isGoogleMapLoaded) ⇒ <code>boolean</code>
        * [.getServerTime()](#Utils.Functions.getServerTime) ⇒ <code>Promise</code>
        * [.getCustomDataInfo(customData, appName, fieldsToSearch)](#Utils.Functions.getCustomDataInfo) ⇒ <code>object</code> \| <code>null</code>
    * [.Validations](#Utils.Validations) : <code>object</code>
        * [.isEmail(email)](#Utils.Validations.isEmail) ⇒ <code>boolean</code>
        * [.isURL(URL)](#Utils.Validations.isURL) ⇒ <code>boolean</code>
        * [.isJSON(json)](#Utils.Validations.isJSON) ⇒ <code>boolean</code>
        * [.isRUT(rut)](#Utils.Validations.isRUT) ⇒ <code>boolean</code>
        * [.isRUC(ruc)](#Utils.Validations.isRUC) ⇒ <code>boolean</code>
        * [.isRFC(RFC)](#Utils.Validations.isRFC) ⇒ <code>boolean</code>

<a name="Utils.Functions"></a>

### Utils.Functions : <code>object</code>
**Kind**: static namespace of [<code>Utils</code>](#Utils)  

* [.Functions](#Utils.Functions) : <code>object</code>
    * [.setCurrency(currency)](#Utils.Functions.setCurrency)
    * [.formatPrice(number, [thousands], [decimals], [length], [currency])](#Utils.Functions.formatPrice) ⇒ <code>string</code>
    * [.sanitizeString(str, [replace])](#Utils.Functions.sanitizeString) ⇒ <code>string</code>
    * [.getResizedImage(src, width, height)](#Utils.Functions.getResizedImage) ⇒ <code>string</code>
    * [.setCookie(cname, cvalue, [exdays], [isdomain])](#Utils.Functions.setCookie) ⇒ <code>void</code>
    * [.getCookie(cname)](#Utils.Functions.getCookie) ⇒ <code>string</code>
    * [.deleteCookie(cname)](#Utils.Functions.deleteCookie) ⇒ <code>void</code>
    * [.addAnimation(name, callback)](#Utils.Functions.addAnimation)
    * [.setStrLength(str, maxLength)](#Utils.Functions.setStrLength) ⇒
    * [.stripHost(URL)](#Utils.Functions.stripHost) ⇒ <code>string</code>
    * [.detectIE()](#Utils.Functions.detectIE) ⇒ <code>string</code> \| <code>false</code>
    * [.calculatePercentDiscount(listprice, bestprice)](#Utils.Functions.calculatePercentDiscount) ⇒ <code>string</code>
    * [.isGoogleMapLoaded()](#Utils.Functions.isGoogleMapLoaded) ⇒ <code>boolean</code>
    * [.getServerTime()](#Utils.Functions.getServerTime) ⇒ <code>Promise</code>
    * [.getCustomDataInfo(customData, appName, fieldsToSearch)](#Utils.Functions.getCustomDataInfo) ⇒ <code>object</code> \| <code>null</code>




<a name="Utils.Functions.setCurrency"></a>

#### Functions.setCurrency(currency)
Sets the currency that will be used by helper functions

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  

| Param | Type | Description |
| --- | --- | --- |
| currency | <code>string</code> | The currency |

<a name="Utils.Functions.formatPrice"></a>

#### Functions.formatPrice(number, [thousands], [decimals], [length], [currency]) ⇒ <code>string</code>
Formats a number

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: <code>string</code> - The formatted price  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| number | <code>number</code> \| <code>string</code> |  | The number to format |
| [thousands] | <code>string</code> | <code>&quot;\&quot;.\&quot;&quot;</code> | thousands delimiter |
| [decimals] | <code>string</code> | <code>&quot;\&quot;,\&quot;&quot;</code> | decimal delimiter |
| [length] | <code>integer</code> | <code>2</code> | length of decimal |
| [currency] | <code>string</code> |  | If true, the currency setted with Utils.setCurrency("$") will be added, if a currency (string) is passed it will use that instead; |

<a name="Utils.Functions.sanitizeString"></a>

#### Functions.sanitizeString(str, [replace]) ⇒ <code>string</code>
Sanitize a string, removing/replacing all special characters and spaces with underscore

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: <code>string</code> - The modified string  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| str | <code>string</code> |  | The string to sanitize |
| [replace] | <code>string</code> | <code>&quot;\&quot;-\&quot;&quot;</code> | The string to replace white spaces with, default "-" |

**Example**  
```js
Utils.sanitizeString("hóla múndo"); //Output "hola-mundo"
```
<a name="Utils.Functions.getResizedImage"></a>

#### Functions.getResizedImage(src, width, height) ⇒ <code>string</code>
Change the width & height from a given VTEX image source

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: <code>string</code> - The resized image source  

| Param | Type | Description |
| --- | --- | --- |
| src | <code>string</code> | The source of the image |
| width | <code>int</code> \| <code>string</code> | The new image with |
| height | <code>int</code> \| <code>string</code> | The new image height |

**Example**  
```js
//Given an image thumb source
Fizzmod.Utils.getResizedImage('http://fizzmod.vteximg.com.br/arquivos/ids/155242-292-292/image.png', 500, 600);
//Output: http://fizzmod.vteximg.com.br/arquivos/ids/155242-500-600/image.png

//Given a full image source
Fizzmod.Utils.getResizedImage('http://fizzmod.vteximg.com.br/arquivos/ids/155242/image.png', 100, 100);
//Output: http://fizzmod.vteximg.com.br/arquivos/ids/155242-100-100/image.png
```
<a name="Utils.Functions.setCookie"></a>

#### Functions.setCookie(cname, cvalue, [exdays], [isdomain]) ⇒ <code>void</code>
set a cookie

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  

| Param | Type | Description |
| --- | --- | --- |
| cname | <code>string</code> | The name of the cookie |
| cvalue | <code>mixed</code> | The value of the cookie, if the value is an object, it will be JSON encoded |
| [exdays] | <code>int</code> | Expiration days, if not set the cookie will last through the session only |
| [isdomain] | <code>bool</code> | Set as domain cookie. (Default false, adding "." before the url.) |

<a name="Utils.Functions.getCookie"></a>

#### Functions.getCookie(cname) ⇒ <code>string</code>
**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: <code>string</code> - - The cookie value  

| Param | Type | Description |
| --- | --- | --- |
| cname | <code>string</code> | The name of the cookie to get |

<a name="Utils.Functions.deleteCookie"></a>

#### Functions.deleteCookie(cname) ⇒ <code>void</code>
Remove cookie in docmuent

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  

| Param | Type | Description |
| --- | --- | --- |
| cname | <code>string</code> | The name of the cookie to delete |

<a name="Utils.Functions.addAnimation"></a>

#### Functions.addAnimation(name, callback)
Add an animation listener for the given animation name

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  

| Param | Type | Description |
| --- | --- | --- |
| name | <code>string</code> | The animation name |
| callback | <code>function</code> | The animation callback |

**Example**  
```js
Fizzmod.Utils.addAnimation('nodeInserted', myFunction);
```
<a name="Utils.Functions.setStrLength"></a>

#### Functions.setStrLength(str, maxLength) ⇒
slice string if string is greater than maxLength

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: new string with three dots  

| Param | Type | Description |
| --- | --- | --- |
| str | <code>string</code> |  |
| maxLength | <code>number</code> | default 27 |

**Example**  
```js
Utils.setStrLength('Fizzmod', 3) // Fizz...
```
<a name="Utils.Functions.stripHost"></a>

#### Functions.stripHost(URL) ⇒ <code>string</code>
Removes the host from an URL

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: <code>string</code> - The modified string  

| Param | Type | Description |
| --- | --- | --- |
| URL | <code>string</code> | The URL |

**Example**  
```js
Utils.stripHost("http://test.vtexcommercestable.com.br/contacto/test"); //  "/contacto/test"
```
<a name="Utils.Functions.detectIE"></a>

#### Functions.detectIE() ⇒ <code>string</code> \| <code>false</code>
Check whether the browser is IE and return the version if so.

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: <code>string</code> \| <code>false</code> - The IE version or false if other browser  
<a name="Utils.Functions.calculatePercentDiscount"></a>

#### Functions.calculatePercentDiscount(listprice, bestprice) ⇒ <code>string</code>
calculates discount percentage between two prices.

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
**Returns**: <code>string</code> - - Return percent discunt rounded in Price  

| Param | Type | Description |
| --- | --- | --- |
| listprice | <code>number</code> | Number price of list |
| bestprice | <code>number</code> | Number price for selling |

**Example**  
```js
Utils.calculatePercentDiscount(100, 50) // 50%
```
<a name="Utils.Functions.isGoogleMapLoaded"></a>

#### Functions.isGoogleMapLoaded() ⇒ <code>boolean</code>
Check if google mao is loadedd

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
<a name="Utils.Functions.getServerTime"></a>

#### Functions.getServerTime() ⇒ <code>Promise</code>
Get the VTEX server time

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  
<a name="Utils.Functions.getCustomDataInfo"></a>

#### Functions.getCustomDataInfo(customData, appName, fieldsToSearch) ⇒ <code>object</code> \| <code>null</code>
Function for get info into especific fields in app into customData

**Kind**: static method of [<code>Functions</code>](#Utils.Functions)  

| Param | Type | Description |
| --- | --- | --- |
| customData | <code>object</code> | customData object into orderForm |
| appName | <code>string</code> | CustomData App name |
| fieldsToSearch | <code>array</code> | Array of string with CustomData app fields names |

<a name="Utils.Validations"></a>

### Utils.Validations : <code>object</code>
**Kind**: static namespace of [<code>Utils</code>](#Utils)  

* [.Validations](#Utils.Validations) : <code>object</code>
    * [.isEmail(email)](#Utils.Validations.isEmail) ⇒ <code>boolean</code>
    * [.isURL(URL)](#Utils.Validations.isURL) ⇒ <code>boolean</code>
    * [.isJSON(json)](#Utils.Validations.isJSON) ⇒ <code>boolean</code>
    * [.isRUT(rut)](#Utils.Validations.isRUT) ⇒ <code>boolean</code>
    * [.isRUC(ruc)](#Utils.Validations.isRUC) ⇒ <code>boolean</code>
    * [.isRFC(RFC)](#Utils.Validations.isRFC) ⇒ <code>boolean</code>

<a name="Utils.Validations.isEmail"></a>

#### Validations.isEmail(email) ⇒ <code>boolean</code>
check if a string is a valid email

**Kind**: static method of [<code>Validations</code>](#Utils.Validations)  

| Param | Type | Description |
| --- | --- | --- |
| email | <code>string</code> | string to check |

<a name="Utils.Validations.isURL"></a>

#### Validations.isURL(URL) ⇒ <code>boolean</code>
Check if a string is a valid URL

**Kind**: static method of [<code>Validations</code>](#Utils.Validations)  

| Param | Type | Description |
| --- | --- | --- |
| URL | <code>string</code> | string to check |

<a name="Utils.Validations.isJSON"></a>

#### Validations.isJSON(json) ⇒ <code>boolean</code>
Check if a string is a valid json

**Kind**: static method of [<code>Validations</code>](#Utils.Validations)  

| Param | Type | Description |
| --- | --- | --- |
| json | <code>string</code> | string to check |

<a name="Utils.Validations.isRUT"></a>

#### Validations.isRUT(rut) ⇒ <code>boolean</code>
Validate RUT (Chile)

**Kind**: static method of [<code>Validations</code>](#Utils.Validations)  

| Param | Type | Description |
| --- | --- | --- |
| rut | <code>string</code> | The rut to validate |

<a name="Utils.Validations.isRUC"></a>

#### Validations.isRUC(ruc) ⇒ <code>boolean</code>
Validate RUC (Perú)

**Kind**: static method of [<code>Validations</code>](#Utils.Validations)  

| Param | Type | Description |
| --- | --- | --- |
| ruc | <code>string</code> | The ruc to validate |

<a name="Utils.Validations.isRFC"></a>

#### Validations.isRFC(RFC) ⇒ <code>boolean</code>
Validate RFC (Mexico)

**Kind**: static method of [<code>Validations</code>](#Utils.Validations)  

| Param | Type | Description |
| --- | --- | --- |
| RFC | <code>string</code> | The RFC to validate |


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.3.0] - 2019-04-31
### Changed
-  change dependency babel/runtine to peer dependency

## [1.2.2] - 2019-04-31
### Added
-  add npm script for build before publish

## [1.2.1] - 2019-04-31
### Fixed
- Fix error in getSreverTime function

## [1.2.0] - 2019-04-31
### Changed
- change repository for package and dependencies for build


test/test.js
import customData from '../src/__mocks__/customData';
import {
	setCurrency,
	formatPrice,
	sanitizeString,
	getResizedImage,
	setStrLength,
	calculatePercentDiscount,
	stripHost,
	isEmail,
	isJSON,
	isRUT,
	isRUC,
	isRFC,
	getCustomDataInfo
} from '../src';

const expect = require('expect.js');

describe('Test functions utils', () => {

	it('check if set correct a currency', () => {
		setCurrency('S/');
		const price = formatPrice(100);
		expect(price).to.be('S/100,00');
	});

	it('check if sanitize correct', () => {
		const string = sanitizeString('Hello world jaja');
		expect(string).to.be('hello-world-jaja');
	});

	it('check if image is resized correctly', () => {
		const url = getResizedImage('http://fizzmod.vteximg.com.br/arquivos/ids/155242-292-292/image.png', 150, 150);
		expect(url).to.be('http://fizzmod.vteximg.com.br/arquivos/ids/155242-150-150/image.png');
	});

	it('check if string it is cutted in the position and added a points', () => {
		const string = setStrLength('Hello', 4);
		expect(string).to.be('Hell...');
	});

	it('check if removed the host of the url', () => {
		const string = stripHost('http://test.vtexcommercestable.com.br/contacto/test');
		expect(string).to.be('/contacto/test');
	});

	it('check if email it is valid', () => {
		const email = isEmail('juancho@test.com');
		expect(email).to.be(true);
	});

	it('check if email it is not valid', () => {
		const email = isEmail('juancho@test');
		expect(email).to.be(false);
	});

	it('check if json it is valid', () => {
		const json = '{"test": "test"}';
		expect(isJSON(json)).to.be(true);
	});

	it('check if json it is not valid', () => {
		const json = '{"test": "test", "asdasd"}';
		expect(isJSON(json)).to.be(false);
	});

	it('check if RUT it is valid', () => {
		const rut = '34297936-K';
		expect(isRUT(rut)).to.be(true);
	});

	it('check if RUT it is not valid', () => {
		const rut = '34.297.936-L';
		expect(isRUT(rut)).to.be(false);
	});

	it('check if RUC it is valid', () => {
		const ruc = '10164181826';
		expect(isRUC(ruc)).to.be(true);
	});

	it('check if RUC it is not valid', () => {
		const ruc = '1016409058';
		expect(isRUC(ruc)).to.be(false);
	});

	it('check if RFC it is valid', () => {
		const RFC = 'CUPU800825569';
		expect(isRFC(RFC)).to.be(true);
	});

	it('check if RFC it is not valid', () => {
		const RFC = 'CUPU80082556';
		expect(isRFC(RFC)).to.be(false);
	});


	it('check if percent discount is correct', () => {
		const lPrice = 100;
		const bPrice = 50;
		expect(calculatePercentDiscount(lPrice, bPrice)).to.be('50%');
	});

	it('check customData get info', () => {
		const data = getCustomDataInfo(customData, 'storepickup', ['data']);
		expect('data' in data).to.be(true);
	});
})


src/functions.js
import axios from 'axios';

/* eslint max-len: ["error", { "ignoreComments": true , "code": 100 }] */
/** @namespace Utils */
/** @namespace Utils.Functions */


const Utils = {
	currency: null,
	animationCallbacks: []
};


/**
 * @function setCurrency
 * @description Sets the currency that will be used by helper functions
 * @memberof Utils.Functions
 * @param {string} currency - The currency
*/
export const setCurrency = (currency) => {
	Utils.currency = currency;
};


/**
 * @function formatPrice
 * @description Formats a number
 * @memberof Utils.Functions
 * @param {number|string} number - The number to format
 * @param {string} [thousands="."] - thousands delimiter
 * @param {string} [decimals=","] - decimal delimiter
 * @param {integer} [length=2] - length of decimal
 * @param {string} [currency] - If true, the currency setted with Utils.setCurrency("$") will be added, if a currency (string) is passed it will use that instead;
 * @returns {string} The formatted price
*/
export const formatPrice = (number, thousands, decimals, length, currency) => {
	const utilsCurrency = Utils.currency ? Utils.currency : '';
	const newCurrency = (currency && typeof currency === 'string') ? currency : utilsCurrency;
	const newLength = typeof length !== 'number' ? 2 : length;

	const re = `\\d(?=(\\d{3})+${newLength > 0 ? '\\D' : '$'})`;
	const regex = new RegExp(re, 'g');
	let newNumber = (number * 1).toFixed(Math.max(0, ~~newLength)).toString(); // eslint-disable-line
	if(typeof newNumber !== 'string')
		newNumber = newNumber.toString();

	newNumber = newNumber.replace('.', (decimals || ','));
	newNumber = newNumber.replace(regex, `$&${thousands || '.'}`);

	if(/€/g.test(newCurrency))
		return newNumber + newCurrency;

	return newCurrency + newNumber;
};


/**
 * @function sanitizeString
 * @description Sanitize a string, removing/replacing all special characters and spaces with underscore
 * @memberof Utils.Functions
 * @param {string} str - The string to sanitize
 * @param {string} [replace="-"] - The string to replace white spaces with, default "-"
 * @returns {string} The modified string
 * @example Utils.sanitizeString("hóla múndo"); //Output "hola-mundo"
*/
export const sanitizeString = (str, rp) => {
	const replace = typeof rp === 'string' ? rp : '-';
	let string = str.toLowerCase();
	string = string.replace(/[\[\]\(\)\-\{\}\^\,]/g, ''); // eslint-disable-line
	string = string.replace(/[àáâãäåª]/g, 'a');
	string = string.replace(/[éèëê]/g, 'e');
	string = string.replace(/[íìïî]/g, 'i');
	string = string.replace(/[óòöô]/g, 'o');
	string = string.replace(/[úùüû]/g, 'u');
	string = string.replace(/[ñ]/g, 'n');
	string = string.replace(/[ç]/g, 'c');
	string = string.replace(/ /g, replace);
	return string;
};


/**
 * @function getResizedImage
 * @description Change the width & height from a given VTEX image source
 * @memberof Utils.Functions
 * @param {string} src - The source of the image
 * @param {int|string} width - The new image with
 * @param {int|string} height - The new image height
 * @returns {string} The resized image source
 * @example
 * //Given an image thumb source
 * Fizzmod.Utils.getResizedImage('http://fizzmod.vteximg.com.br/arquivos/ids/155242-292-292/image.png', 500, 600);
 * //Output: http://fizzmod.vteximg.com.br/arquivos/ids/155242-500-600/image.png
 *
 * //Given a full image source
 * Fizzmod.Utils.getResizedImage('http://fizzmod.vteximg.com.br/arquivos/ids/155242/image.png', 100, 100);
 * //Output: http://fizzmod.vteximg.com.br/arquivos/ids/155242-100-100/image.png
*/
export const getResizedImage = (src, width, height) => {
	if(width === undefined || height === undefined || typeof src !== 'string')
		return src;

	const pattern = /(?:ids\/[0-9]+)-([0-9]+)-([0-9]+)\//;

	const newSrc = src.replace(pattern, (match, matchedWidth, matchedHeight) => (
		match.replace(`-${matchedWidth}-${matchedHeight}`, `-${width}-${height}`)
	));

	return newSrc.replace(/(ids\/[0-9]+)\//, `$1-${width}-${height}/`);
};


/**
 * set a cookie
 * @function setCookie
 * @memberof Utils.Functions
 * @param {string} cname - The name of the cookie
 * @param {mixed} cvalue - The value of the cookie, if the value is an object, it will be JSON encoded
 * @param {int} [exdays] - Expiration days, if not set the cookie will last through the session only
 * @param {bool} [isdomain] - Set as domain cookie. (Default false, adding "." before the url.)
 * @returns {void}
*/
export const setCookie = (cname, cvalue, exdays, isdomain = false) => {
	let expires = '';
	const value = typeof cvalue === 'object' ? JSON.stringify(cvalue) : cvalue;

	if(!Number.isNaN(exdays)) {
		const d = new Date();
		d.setTime(d.getTime() + (exdays * 24 * 60 * 60 * 1000));
		expires = `expires=${d.toGMTString()};`;
	}

	if(isdomain)
		document.cookie = `${cname}=${value}; ${expires}path=/`;
	else
		document.cookie = `${cname}=${value};domain=.${window.location.host}; ${expires}path=/`;
};


/**
 * @function getCookie
 * @memberof Utils.Functions
 * @param {string} cname - The name of the cookie to get
 * @returns {string} - The cookie value
*/
export const getCookie = (cname) => {
	const value = `; ${document.cookie}`;
	const parts = value.split(`; ${cname}=`);
	if(parts.length === 2)
		return parts.pop().split(';').shift();
	return '';
};


/**
 * @function deleteCookie
 * @description Remove cookie in docmuent
 * @memberof Utils.Functions
 * @param {string} cname - The name of the cookie to delete
 * @returns {void}
*/
export const deleteCookie = cname => setCookie(cname, '', -1);


/**
* @function addAnimation
* @description Add an animation listener for the given animation name
* @memberof Utils.Functions
* @param {string} name - The animation name
* @param {function} callback - The animation callback
* @example
* Fizzmod.Utils.addAnimation('nodeInserted', myFunction);
*/
export const addAnimation = (name, callback) => {
	const { animationCallbacks } = Utils;
	const listenerCallback = (e) => {
		if(e.animationName in animationCallbacks) {
			// eslint-disable-next-line
			for(let i = 0, animationLength = animationCallbacks[e.animationName].length; i < animationLength; i += 1)
				animationCallbacks[e.animationName][i].call(null, e);
		}
	};

	if(!Object.keys(animationCallbacks).length) {
		document.addEventListener('animationstart', listenerCallback);
		document.addEventListener('webkitAnimationStart', listenerCallback);
		document.addEventListener('MSAnimationStart', listenerCallback);
		document.addEventListener('oAnimationStart', listenerCallback);
		document.addEventListener('mozAnimationStart', listenerCallback);
	}
	if(!(animationCallbacks[name] instanceof Array))
		animationCallbacks[name] = [];

	animationCallbacks[name].push(callback);
};


/**
 * @function setStrLength
 * @memberof Utils.Functions
 * @param {string} str
 * @param {number} maxLength - default 27
 * @description slice string if string is greater than maxLength
 * @returns new string with three dots
 * @example Utils.setStrLength('Fizzmod', 3) // Fizz...
*/
export const setStrLength = (str, maxLength = 27) => {
	if(!str)
		return;
	return (str.length >= maxLength) ? `${str.slice(0, maxLength)}...` : str;
};


/**
 * @function stripHost
 * @description Removes the host from an URL
 * @memberof Utils.Functions
 * @param {string} URL - The URL
 * @returns {string} The modified string
 * @example Utils.stripHost("http://test.vtexcommercestable.com.br/contacto/test"); //  "/contacto/test"
 */
export const stripHost = url => url.toString().replace(/https?:\/\/.*?\//i, '/');


/**
 * @function detectIE
 * @description Check whether the browser is IE and return the version if so.
 * @memberof Utils.Functions
 * @returns {string|false} The IE version or false if other browser
*/
export const detectIE = () => {
	const ua = window.navigator.userAgent;

	const msie = ua.indexOf('MSIE ');
	if(msie > 0) {
		// IE 10 or older => return version number
		return parseInt(ua.substring(msie + 5, ua.indexOf('.', msie)), 10);
	}

	const trident = ua.indexOf('Trident/');
	if(trident > 0) {
		// IE 11 => return version number
		const rv = ua.indexOf('rv:');
		return parseInt(ua.substring(rv + 3, ua.indexOf('.', rv)), 10);
	}

	const edge = ua.indexOf('Edge/');
	if(edge > 0) {
		// IE 12 => return version number
		return parseInt(ua.substring(edge + 5, ua.indexOf('.', edge)), 10);
	}

	// other browser
	return false;
};

/**
 * @function calculatePercentDiscount
 * @memberof Utils.Functions 
 * @description calculates discount percentage between two prices.
 * @param {number} listprice - Number price of list
 * @param {number} bestprice - Number price for selling
 * @returns {string} - Return percent discunt rounded in Price
 * @example Utils.calculatePercentDiscount(100, 50) // 50%
*/
export const calculatePercentDiscount = (listprice, bestprice) => {
	const lPrice = parseInt(listprice, 10);
	const bPrice = parseInt(bestprice, 10);
	const percentDiscount = ((lPrice - bPrice) / lPrice) * 100;
	return `${Math.round(percentDiscount)}%`;
};

/**
 * @function isGoogleMapLoaded
 * @memberof Utils.Functions 
 * @description Check if google mao is loadedd
 * @returns {boolean}
*/
export const isGoogleMapLoaded = () => (
	typeof window.google !== 'undefined' && typeof window.google.maps !== 'undefined'
);

/**
 * @description Get the VTEX server time
 * @memberof Utils.Functions 
 * @function getServerTime
 * @returns {Promise}
*/
export const getServerTime = async() => {
	const { data } = await axios.get('/no-cache/HoraAtualServidor.aspx');

	const monthBr = ['jan', 'fev', 'mar', 'abr', 'mai', 'jun', 'jul', 'ago', 'set', 'out', 'nov', 'dez'];

	const time = data.match(/([0-9]+):([0-5][0-9]):([0-5][0-9])/)[0];
	let day = parseInt(data.match(/[a-z]{3} ([0-9]{1,2})/)[1], 10);
	let month = monthBr.indexOf(data.match(/[a-z]{3}/)[0]) + 1;
	const year = parseInt(data.match(/[0-9]{4}/)[0], 10);

	if(day < 10)
		day = `0${day}`;

	if(month < 10)
		month = `0${month}`;

	return new Date(`${year}/${month}/${day} ${time}`);
};

/**
 * @function getCustomDataInfo
 * @memberof Utils.Functions
 * @description Function for get info into especific fields in app into customData
 * @param {object} customData - customData object into orderForm
 * @param {string} appName - CustomData App name
 * @param {array} fieldsToSearch - Array of string with CustomData app fields names
 * @returns {object|null}
*/
export const getCustomDataInfo = (customData, appName, fieldsToSearch) => {
	try {

		if(!(customData && customData instanceof Object && !(customData instanceof Array)))
			throw 'Customdata invalid';
		if(!(appName && typeof appName === 'string'))
			throw 'appName invalid';
		if(!(fieldsToSearch && fieldsToSearch instanceof Array && fieldsToSearch.every(e => typeof e === 'string')))
			throw 'fieldsToSearch invalid';

		if('customApps' in customData) {
			const { customApps } = customData;
			if(customApps.length) {
				const objSearch = customApps.find(app => appName === app.id);
				
				if(!objSearch)
					throw `appName '${appName}' not found`;

				const { fields } = objSearch;
				
				if(fieldsToSearch instanceof Array) {
					const data = fieldsToSearch.reduce((acccum, fieldName) => {
						const accumulator = Object.assign({}, acccum);

						if(fields[fieldName])
							accumulator[fieldName] = fields[fieldName];

						return accumulator;
					}, {});

					if(!Object.keys(data).length)
						throw `not found any 'fieldsToSearch'`;

					return data;
				}
			}
		}
	} catch(error) {
		console.error(new Error(error));
		return null;
	}
};

